### PR TITLE
Add thickness and event slices to table

### LINKED ISSUE #21260: Add thickness column to SANS GUI
The thickness is a common cause of error among users so should have it's own prominent column in the table.

### PR DESCRIPTION
**Pull request by Matthew-Andrew**
This PR creates a new column in the batch reduction table in which the sample thickness will be displayed and can be changed. 

This PR also added EventSlices as an option in the options column in the table.

**To test:**
- Get some sans data the loqdemo folder from the ISIS sample data here downloads.mantidproject.org will work.
- Open the new sans GUI, Load MaskFile.txt as the user file and batch_mode_reduction.csv as the batch file.
- Confirm that the thickness column is populated. 
- Change the thickness column and then go to settings->diagnostics. 
- Click on update rows
- Check that in scale the thickness is set to the thickness you have entered and that the thickness_from_file value equals the original value.

- To test the event slicing go back to the runs tab. In the options column type EventSlices='1:1:5,3-7'
- Go to settings->diagnostics and click update rows again
- in the slice settings confirm that end_time = [2,3,4,5,7] and start_time = [1,2,3,4,3]

Fixes #21260 and #21255 . 

**Release Notes** 

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: scripts/SANS/sans/gui_logic/models/table_model.py [hunk 1]
```diff
@@ -117,17 +118,20 @@ def _parse_string(options_column_string):
...
         """
-        parsed = {}
         # Remove all white space
+        parsed = {}
         options_column_string_no_whitespace = "".join(options_column_string.split())
+        options_column_string_no_whitespace = options_column_string_no_whitespace.replace('"', '')
+        options_column_string_no_whitespace = options_column_string_no_whitespace.replace("'", '')
         if not options_column_string_no_whitespace:
             return parsed
+        pat = re.compile(r'''([^,=]+)=*((?:[^,=]+(?:,|$))*)''')
```
Comment (AnthonyLim23) [inline scripts/SANS/sans/gui_logic/models/table_model.py]: sorry, what is this doing?

File: scripts/SANS/sans/gui_logic/models/table_model.py [hunk 2]
```diff
@@ -117,17 +118,20 @@ def _parse_string(options_column_string):
...
+
+        for k, v in pat.findall(options_column_string_no_whitespace):
+            if v.endswith(','):
+                v=v[:-1]
+            parsed.update({k:v})
```
Comment (AnthonyLim23) [inline scripts/SANS/sans/gui_logic/models/table_model.py]: why did you change this code? The old version seems more readable.

File: scripts/SANS/sans/gui_logic/presenter/gui_state_director.py [hunk 3]
```diff
@@ -53,6 +53,10 @@ def create_state(self, row):
         if output_name:
             state_gui_model.output_name = output_name
 
+        if table_index_model.sample_thickness:
+            sample_thickness = float(table_index_model.sample_thickness)
+            state_gui_model.sample_thickness = sample_thickness
```
Comment (AnthonyLim23) [inline scripts/SANS/sans/gui_logic/presenter/gui_state_director.py]: could we combine these two lines into one?

File: scripts/test/SANS/gui_logic/main_presenter_test.py [hunk 4]
```diff
@@ -21,7 +21,7 @@ def test_that_the_white_list_is_correct(self):
         presenter = MainPresenter(SANSFacility.ISIS)
         self.assertTrue(presenter.get_number_of_white_list_items() == 0)
         white_list = presenter.get_white_list(show_periods=True)
-        self.assertTrue(presenter.get_number_of_white_list_items() == 19)
+        self.assertEqual(presenter.get_number_of_white_list_items(), 20)
```
Comment (AnthonyLim23) [inline scripts/test/SANS/gui_logic/main_presenter_test.py]: is the increase to 20 just because you have added a row?

File: scripts/test/SANS/gui_logic/run_tab_presenter_test.py [hunk 5]
```diff
@@ -138,7 +138,7 @@ def test_that_will_load_user_file(self):
         # Assert certain function calls
         self.assertTrue(view.get_user_file_path.call_count == 3)
         self.assertTrue(view.get_batch_file_path.call_count == 2)  # called twice for the sub presenter updates (masking table and settings diagnostic tab)  # noqa
-        self.assertTrue(view.get_cell.call_count == 64)
+        self.assertEqual(view.get_cell.call_count, 68)
```
Comment (AnthonyLim23) [inline scripts/test/SANS/gui_logic/run_tab_presenter_test.py]: why has this increase?

Comment (Mantid-Matthew) [reply scripts/test/SANS/gui_logic/run_tab_presenter_test.py]: This has increased because we have added a new column to the table so as we loop through the table getting values from cells we do so four more times overall.

File: scripts/test/SANS/gui_logic/main_presenter_test.py (hunk 4 above)
Comment (Mantid-Matthew) [reply scripts/test/SANS/gui_logic/main_presenter_test.py]: This is because I added a new column yes.

File: scripts/SANS/sans/gui_logic/sans_data_processor_gui_algorithm.py [hunk 6]
```diff
@@ -45,7 +45,9 @@ def create_option_column_properties():
                                   property_type=float),
              algorithm_list_entry(column_name="",
                                   algorithm_property="EventSlices",
-                                  description='The event slices to take',
+                                  description='The event slices to reduce. The format is the same as for the event slices'
+                                              ' box in settings however if a comma separated list is given '
```
Comment (AnthonyLim23) [inline scripts/SANS/sans/gui_logic/sans_data_processor_gui_algorithm.py]: if I am being fussy then you should have a comma before however. but it doesn't matter

**Comment by AnthonyLim23** [comment]
I have loaded the files stated but the thickness col is not populated

image

**Review by AnthonyLim23 — APPROVED** [review]
works as described and code looks good

**Comment by martyngigg** [comment]
OSX failure is unrelated.
